pkg/common/cluster_state.go

```diff
@@ -0,0 +1,43 @@
+package common
+
+import (
+	"context"
+	kc "github.com/keycloak/keycloak-operator/pkg/apis/keycloak/v1alpha1"
+	"github.com/keycloak/keycloak-operator/pkg/model/keycloak"
+	v1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// The desired cluster state is defined by a list of actions that have to be run to
+// get from the current state to the desired state
+type DesiredClusterState []ClusterAction
+
+type ClusterState struct {
+	KeycloakService *v1.Service
+}
+
+func NewClusterState() *ClusterState {
+	return &ClusterState{
+		KeycloakService: nil,
+	}
+}
+
+func (i *ClusterState) Read(cr *kc.Keycloak, controllerClient client.Client) {
+	i.readKeycloakServiceCurrentState(cr, controllerClient)
+}
+
+// Keycloak service
+func (i *ClusterState) readKeycloakServiceCurrentState(cr *kc.Keycloak, controllerClient client.Client) {
+	keycloakService := keycloak.Service(cr)
+
+	selector := client.ObjectKey{
+		Name:      keycloakService.Name,
+		Namespace: keycloakService.Namespace,
+	}
+	err := controllerClient.Get(context.TODO(), selector, keycloakService)
+	if err != nil {
```

Comment:
The way I am reading this error handling is if there is no error returned from the `Get`, the current cluster state for the keycloakservice will be set to `nil`, if there is an error, it set the cluster state to the keycloakService object. 

This should be reversed and also handle the difference between an error and when the resource doesn't exist. 

I haven't been able to find much from the docs on what is returned when the resource doesn't exist. https://godoc.org/sigs.k8s.io/controller-runtime/pkg/client#DelegatingReader.Get

Reply:
I looked through the code and it indeed seems fine :fearful: 

The `Get` method tries to obtain an objects from a cache (see [here](https://github.com/keycloak/keycloak-operator/blob/4dfd3d9d1b8e6011c0c50ab7b4b08f4804046677/vendor/sigs.k8s.io/controller-runtime/pkg/cache/internal/cache_reader.go#L53)). If the resources, doesn't exist - it returns an error. 

So in short, if there's an error (note the `!=` in the code), the object doesn't exist. Therefore, we set the `i.KeycloakService = nil`.

Reply:
Thanks @slaskawi . You are correct, I was reading this wrong. So in the case here where an error is returned, we need to handle legitimate errors as well. We should differentiate Not Found errors which will set it to nil and all other errors should return an error code. This can be done by checking the error type is `NewNotFound`. https://github.com/keycloak/keycloak-operator/blob/4dfd3d9d1b8e6011c0c50ab7b4b08f4804046677/vendor/sigs.k8s.io/controller-runtime/pkg/cache/internal/cache_reader.go#L61

Reply:
There is a helper method for that in the `errors` package (IsNotFoundError)

Reply:
Implemented.